### PR TITLE
Expose additional functionality in collections and cross-memory

### PR DESCRIPTION
**Pull request by ifakhrutdinov**
#### Overview
This pull request add the following features:
* Ability to use the lock-free queue intrusively which allows a more flexible storage management on the user's side
* Functions to copy to/from foreign address space using destination/source keys and ALETs

### CODE REVIEW

File: h/collections.h [hunk 1]
```diff
@@ -259,9 +259,52 @@ ZOWE_PRAGMA_PACK_RESET
...
+ */
 void destroyQueue(Queue *q) QueueAmode64;
+
+#ifdef __ZOWE_OS_ZOS
+
+/**
+ * Enqueue a queue element.
+ * @param q Queue to which the new element will be enqueued.
+ * @param newElement Queue element to be enqueued. WARINIG: it must reside on
+ * a double word boundary, otherwise a specification exception (S0C6 ABEND)
```
Comment (daveyc) [inline h/collections.h]: I'm not sure if double word alignment is required (ABEND0C6) or recommended for performance reasons. Maybe Greg can comment?

File: c/collections.c [hunk 2]
```diff
@@ -1343,11 +1350,17 @@ void *qRemove(Queue *q){
...
+
   void *result = NULL;
-  if (currentHead)
-  {
-    result = currentHead->data;
-    safeFree31((char*)currentHead,sizeof(QueueElement));
+  QueueElement *element = qDequeue(q);
+  if (element) {
+    result = element->data;
+    safeFree((char *)element, sizeof(QueueElement));
```
Comment (daveyc) [inline c/collections.c]: Nitpick! This cast to (char *) bothers me and should be fixed in the future. It's not a breaking change and removes an irritating requirement for an unnecessary cast.

File: c/collections.c [hunk 3]
```diff
@@ -1237,7 +1234,17 @@ void qInsert(Queue *q, void *newData){
   return;
 }
 
-void *qRemove(Queue *q){
+void qInsert(Queue *q, void *newData) {
```
Comment (daveyc) [inline c/collections.c]: newData should be const

Comment (ifakhrutdinov) [reply c/collections.c]: I'd need to think this through. If we make this const, we need to make the field const and `qRemove` will need to return a const value which will break existing code. I don't see how we can solve this without casting at some point.

File: c/collections.c (hunk 2 above)
Comment (ifakhrutdinov) [reply c/collections.c]: I agree.

File: c/collections.c (hunk 3 above)
Comment (daveyc) [reply c/collections.c]: I think a cast is fine. I would  cast to (void *) when assigning the data element and leave everything else as is. `qRemove` should remain as is. That's pretty standard. Making newData const signifies the intent of the function in that it doesn't mutate newData.

**Pull request merged**